### PR TITLE
Add nested `PackgeIdentity.Scope` and `PackageIdentity.Name` types

### PR DESCRIPTION
**Pull request by mattt**
Adds package identity scope and name types as defined by SE-0292 (according to updated naming rules proposed by #3624).

### Motivation:

Implement [SE-0292](https://github.com/apple/swift-evolution/blob/main/proposals/0292-package-registry-service.md).

### Modifications:

- Adds a new `PackgeIdentity.Scope` type and related tests
- Adds a new `PackageIdentity.Name` type and related tests

### Result:

The types added by this PR are tested, but not otherwise used by other APIs, so this change has no immediate effect.

### CODE REVIEW

**Comment by mattt** [comment]
@swift-ci Please smoke test

**Comment by mattt** [comment]
@tomerd I'm currently blocked on #3640 until I have a better idea of how we're going to adapt the current package identity code to support local, repository, and registry packages. This PR attempts to lay some groundwork towards that end goal, whatever that looks like.

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
cc @yim-lee for another set of eyes on the validation logic

**Comment by neonichu** [comment]
I think the rough idea would be:
- Local dependencies will continue to use the last-path component as identity and will require explicit manual configuration by the user to give them a scope/name based identity for the purpose of overrides
- URL based dependencies can query the registry to get a scope/name based identity if available and there would also be explicit manual configuration by the user, similar to local dependencies. If none of this is present, they'd continue to use legacy identity
- Registry packages always have a scope/name based identity

I think this implies that prior to #3640 we have to implement a new, explicit override mechanism.

File: Sources/PackageModel/PackageIdentity.swift
Comment (yim-lee) [inline Sources/PackageModel/PackageIdentity.swift:102]: Can we use regex patterns to do the validation?

Comment (mattt) [reply Sources/PackageModel/PackageIdentity.swift:102]: We could, but I can think of a few drawbacks to that approach:
- Less precise errors (a regex match is all or nothing, so you can't know why it's invalid)
- Somewhat slower performance (a regex engine is slower than what we're doing here)
- More coupling (it creates a dependency on a regex engine, which could complicate cross-platform deployment)

On the plus side, a regex is a more obvious representation than string manipulation code. But with comprehensive testing, that's probably less of a concern.

**Comment by mattt** [comment]
> I think the rough idea would be:
> 
> * Local dependencies will continue to use the last-path component as identity and will require explicit manual configuration by the user to give them a scope/name based identity for the purpose of overrides
> * URL based dependencies can query the registry to get a scope/name based identity if available and there would also be explicit manual configuration by the user, similar to local dependencies. If none of this is present, they'd continue to use legacy identity
> * Registry packages always have a scope/name based identity

@neonichu That sounds about right to me. 

Implementation-wise, one approach to help us navigate the transition would be to change  `PackageIdentity` to an enumeration (we _should_ be able to do that while keeping the existing APIs stable). Instead of initializers setting a stored `description` property, the `description` becomes a computed property based on 1) the enumeration case, and 2) the chosen resolution logic (i.e. "legacy" vs. "canonical"). Something like this:

```swift
public enum PackageIdentity: Hashable {
   case synthesized(name: String)
   case local(path: AbsolutePath)
   case repository(url: String)
   case registry(scope: Scope, name: Name)

   /* ... */
}
```

File: Sources/PackageModel/PackageIdentity.swift
Comment (yim-lee) [inline Sources/PackageModel/PackageIdentity.swift:198]: Double-`if` unintended I believe?

Comment (mattt) [reply Sources/PackageModel/PackageIdentity.swift:198]: Yes, correct. Good catch!

**Comment by mattt** [comment]
@swift-ci Please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Pull request merged**
